File: jack_knife.cpp
// Jack knife error
#include <iostream>
#include <cmath>
#include<fstream>
#include<vector>

double func(double x){
	return pow(x, 2);
}


int main(){
	std::ifstream inputFile("input.txt");
	std::vector<double> data;
    double number;

	while (inputFile >> number){
		data.push_back(number);
	}

	inputFile.close();
	int data_size = data.size();
	std::cout << data_size << "data read" << std::endl;

    //Average
	double sum = 0e0;
	for (double num : data){
		sum += func(num);
	}
    std::cout << "Average:" <<  sum / (double)(data_size) << std::endl;

	//Error
	int bin_size = 1;
	int sample_num = (int) data_size / bin_size;
	double sample[sample_num];
	double sum_sample = 0e0;

    for (int i=0; i < sample_num; i++){
		double sum_jack;
		sum_jack = 0e0;

        for (int k=0; k < data_size; k++){
            if (k < bin_size * i || bin_size * (i+1) - 1 < k){
				sum_jack += func(data[k]);
			}
		}
		sample[i] = sum_jack / (double)(data_size - bin_size);
		sum_sample += sample[i];
	}

	double sample_aver = sum_sample / (double)(sample_num);
	double sum_error = 0e0;
    for (int i=0; i < sample_num; i++){
		sum_error += pow(sample[i] - sample_aver, 2);
	}

	std::cout << "Error:" << sqrt(sum_error * (double)(sample_num-1) / (double)(sample_num)) << std::endl;

	return 0;
}

File: jackknife.py
# Calculate jack knife error
import numpy as np


# Any function
def func(x):
    return sum(x)


def jack_knife(data, bin_size):
    sample_num = int(len(data) / bin_size)
    data_jack = [0] * (len(data) - bin_size)
    sample = [0] * sample_num

    for i in range(sample_num):
        for j in range(len(data)):
            k = 0
            if j < i * bin_size or (i + 1) * bin_size <= j:
                data_jack[k] += data[j]
                k += 1
        sample[i] = func(data_jack)

    sample_mean = np.mean(sample)
    var = 0
    for i in range(sample_num):
        var += (sample[i] - sample_mean) ** 2
    error = np.sqrt(var * (sample_num - 1) / sample_num)
    return error


# If you have a data file
# data = np.loadtxt('data.txt')
data = [1, 2, 3, 4, 5, 6]
# len(data) / bin_size must be an integer
bin_size = 2

error = jack_knife(data, bin_size)
print(f"Jack knife error:{error}")
